Large refactor of the toolbox

RoboErikG opened the pull request:
This refactor is still a WIP. It still needs tests and additional cleanup, but the code seems to be in the right place and works in all demos.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/google/blockly-android/507)
<!-- Reviewable:end -->

AnmAtAnm commented:
Reviewed 12 of 38 files at r1.
Review status: 12 of 38 files reviewed at latest revision, 18 unresolved discussions.

---

*[blocklydemo/src/main/java/com/google/blockly/android/demo/StylesActivity.java, line 32 at r1](https://reviewable.io:443/reviews/google/blockly-android/507#-KbWaAbbKSw8hdznfMQ1:-KbWaAbbKSw8hdznfMQ2:b-fg8zdv) ([raw file](https://github.com/google/blockly-android/blob/4ed5c0a35d55fd7c462c8b798f6863f293b4af8e/blocklydemo/src/main/java/com/google/blockly/android/demo/StylesActivity.java#L32)):*
> ```Java
>  * Basic implementation of AbstractBlocklyActivity that demonstrates applying styles. This demo
>  * uses an Activity style defined in the manifest XML. Alternatively, Activities can override
>  * {@link #getStyleResId()} to change the style used in {@link #onCreate(Bundle)}.
> ```

Fixed fully in #508. Revert if you want a simpler merge.

---

*[blocklydemo/src/main/java/com/google/blockly/android/demo/StylesActivity.java, line 79 at r1](https://reviewable.io:443/reviews/google/blockly-android/507#-KbWau1sPnoxCyAndsKe:-KbWau1sPnoxCyAndsKf:b-l4yjhg) ([raw file](https://github.com/google/blockly-android/blob/4ed5c0a35d55fd7c462c8b798f6863f293b4af8e/blocklydemo/src/main/java/com/google/blockly/android/demo/StylesActivity.java#L79)):*
> ```Java
>     @Override
>     protected View onCreateContentView(int parentId) {
>         View root = getLayoutInflater().inflate(R.layout.styled_content, null);
> ```

Is there a particular reason to add this to the styling demo?  If you're just demoing a different layout, can we use a new activity, named to describe why it is different.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java, line 65 at r1](https://reviewable.io:443/reviews/google/blockly-android/507#-KbWwBzysr-401fTZgRs:-KbWwBzysr-401fTZgRt:bb7462z) ([raw file](https://github.com/google/blockly-android/blob/4ed5c0a35d55fd7c462c8b798f6863f293b4af8e/blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java#L65)):*
> ```Java
> 
> /**
>  * TODO UPDATE
> ```

TODO

---

*[blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java, line 116 at r1](https://reviewable.io:443/reviews/google/blockly-android/507#-KbWwbiVeZHd8cOrWT68:-KbWwbiVeZHd8cOrWT69:b-9x3bkz) ([raw file](https://github.com/google/blockly-android/blob/4ed5c0a35d55fd7c462c8b798f6863f293b4af8e/blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java#L116)):*
> ```Java
>     protected BlockClipDataHelper mClipDataHelper;
>     protected WorkspaceFragment mWorkspaceFragment;
>     protected FlyoutFragment mFlyoutFragment;
> ```

mToolboxFlyoutFragment, or just mToolboxFlyout, to distinguish it from the other FlyoutFragment (mTrashFragment).

---

*[blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java, line 853 at r1](https://reviewable.io:443/reviews/google/blockly-android/507#-KbWxXwPAipvlc5ZcvhV:-KbWxXwPAipvlc5ZcvhW:byco7bn) ([raw file](https://github.com/google/blockly-android/blob/4ed5c0a35d55fd7c462c8b798f6863f293b4af8e/blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java#L853)):*
> ```Java
>      *         Otherwise false.
>      */
>     protected boolean onBackToCloseTrash() {
> ```

Rename as onBackToCloseFlyouts()

---

*[blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java, line 319 at r2](https://reviewable.io:443/reviews/google/blockly-android/507#-KbkJp1FkzylP_zKlDx7:-KbkJp1FkzylP_zKlDx8:bb7462z) ([raw file](https://github.com/google/blockly-android/blob/377075ade76b36352eaba49df5efe1cf25426832/blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java#L319)):*
> ```Java
>                 .setToolboxConfigurationAsset(getToolboxContentsXmlPath())
>                 .setTrashFragment(mTrashFragment)
>                 .setToolboxFragment(mFlyoutFragment, mCategoryFragment); // TODO flyout and categories
> ```

TODO

---

*[blocklylib-core/src/main/java/com/google/blockly/android/CategoryFragment.java, line 42 at r1](https://reviewable.io:443/reviews/google/blockly-android/507#-KbWxkGi4dmxwCEgeut4:-KbWxkGi4dmxwCEgeut5:b7tjmd7) ([raw file](https://github.com/google/blockly-android/blob/4ed5c0a35d55fd7c462c8b798f6863f293b4af8e/blocklylib-core/src/main/java/com/google/blockly/android/CategoryFragment.java#L42)):*
> ```Java
>  */
> 
> public class CategoryFragment extends Fragment {
> ```

CategoriesFragment, or CategorySelectorFragment?
TODO(###): Define an interface that app developers can implement their own UI.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/FlyoutFragment.java, line 242 at r1](https://reviewable.io:443/reviews/google/blockly-android/507#-KbXWySgy0T55SALVtb6:-KbXWySgy0T55SALVtb7:b-pwq767) ([raw file](https://github.com/google/blockly-android/blob/4ed5c0a35d55fd7c462c8b798f6863f293b4af8e/blocklylib-core/src/main/java/com/google/blockly/android/FlyoutFragment.java#L242)):*
> ```Java
>         // While it would be nice to see the customized LinearLayoutManager here, its use reveals
>         // a double bind error in RecyclerView, that crashes when binding a BlockGroup that has an
>         // existing parent.
> ```

This comment is obsolete.  Please remove.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/FlyoutFragment.java, line 131 at r2](https://reviewable.io:443/reviews/google/blockly-android/507#-KbqUQhN2hhfn8ebRrjY:-KbqUQhN2hhfn8ebRrjZ:bxyyh60) ([raw file](https://github.com/google/blockly-android/blob/377075ade76b36352eaba49df5efe1cf25426832/blocklylib-core/src/main/java/com/google/blockly/android/FlyoutFragment.java#L131)):*
> ```Java
>         mFlyoutView = (FlyoutView) inflater.inflate(layout, null);
>         mFlyoutView.setLayoutManager(createLinearLayoutManager());
>         return mFlyoutView;
> ```

Maybe not now, but...
I think this need a FrameView or similar, whose contents can be replaced with the correct view in the case of `custom` categories.
At least add a TODO.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyController.java, line 262 at r2](https://reviewable.io:443/reviews/google/blockly-android/507#-KbkKODwAPWxlASsemX8:-KbkKODwAPWxlASsemX9:b-85gfpc) ([raw file](https://github.com/google/blockly-android/blob/377075ade76b36352eaba49df5efe1cf25426832/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyController.java#L262)):*
> ```Java
>      * @param categoryFragment Optional fragment for displaying toolbox categories.
>      */
>     public void setToolboxFragments(FlyoutFragment flyoutFragment,
> ```

Based on a comment above, I think you were considering breaking this into two set methods, which I support.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/FlyoutController.java, line 33 at r2](https://reviewable.io:443/reviews/google/blockly-android/507#-KbkL9mJeDjBPaCMXM57:-KbkL9mKbIjcUA8bGci5:b-iuhzf0) ([raw file](https://github.com/google/blockly-android/blob/377075ade76b36352eaba49df5efe1cf25426832/blocklylib-core/src/main/java/com/google/blockly/android/control/FlyoutController.java#L33)):*
> ```Java
> 
> /**
>  * Provides control logic for the toolbox and trash in a workspace.
> ```

Add: Ensures the toolbox and trash block-list flyouts are populated, opened, and closed in coordination.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/FlyoutController.java, line 173 at r2](https://reviewable.io:443/reviews/google/blockly-android/507#-KbqJWHmAwTf_xmAg1Q6:-KbqJWHmAwTf_xmAg1Q7:bh58xg6) ([raw file](https://github.com/google/blockly-android/blob/377075ade76b36352eaba49df5efe1cf25426832/blocklylib-core/src/main/java/com/google/blockly/android/control/FlyoutController.java#L173)):*
> ```Java
>                 mToolboxFlyout.setCurrentCategory(null);
>             }
>             // TODO reset fragments
> ```

TODO... Not sure if you intend to address this in this pass.  If not, please clarify "reset fragments"

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/FlyoutController.java, line 233 at r2](https://reviewable.io:443/reviews/google/blockly-android/507#-KbkSL9t_wvS4MrIPH18:-KbkSL9t_wvS4MrIPH19:b-cg81ex) ([raw file](https://github.com/google/blockly-android/blob/377075ade76b36352eaba49df5efe1cf25426832/blocklylib-core/src/main/java/com/google/blockly/android/control/FlyoutController.java#L233)):*
> ```Java
>         trashIcon.setOnDragListener(new OnDragToTrashListener(mController));
>         if (mTrashFlyout.isCloseable()) {
>             closeTrash();
> ```

I don't think registering the trashIcon should change the flyout state.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/FlyoutView.java, line 34 at r2](https://reviewable.io:443/reviews/google/blockly-android/507#-KbqT5fXMMgcCspdwnHQ:-KbqT5fXMMgcCspdwnHR:bzctmqb) ([raw file](https://github.com/google/blockly-android/blob/377075ade76b36352eaba49df5efe1cf25426832/blocklylib-core/src/main/java/com/google/blockly/android/ui/FlyoutView.java#L34)):*
> ```Java
>  * View for displaying a list of blocks, labels, and buttons.
>  */
> public class FlyoutView extends RelativeLayout {
> ```

Extract RecyclerView initialization and adapter into a helper or factory method. Maybe into the FlyoutController. Aim to make this class as minimal as possible, possibly with the expectation that it may go away / become a simple RelativeLayout once #503 is resolved.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/FlyoutView.java, line 79 at r2](https://reviewable.io:443/reviews/google/blockly-android/507#-KbqL9da10_f5MyLbdUz:-KbqL9da10_f5MyLbdV-:bnh3aku) ([raw file](https://github.com/google/blockly-android/blob/377075ade76b36352eaba49df5efe1cf25426832/blocklylib-core/src/main/java/com/google/blockly/android/ui/FlyoutView.java#L79)):*
> ```Java
> 
>         if (mActionButton != null) {
>             // TODO (#503): Refactor the action button to be part of the FlyoutCategory.
> ```

I'm assuming moving this variable-specific code into a variable class

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/FlyoutView.java, line 81 at r2](https://reviewable.io:443/reviews/google/blockly-android/507#-KbqKCwSJz4yOfohBUCZ:-KbqKCwSJz4yOfohBUC_:b5az9v9) ([raw file](https://github.com/google/blockly-android/blob/377075ade76b36352eaba49df5efe1cf25426832/blocklylib-core/src/main/java/com/google/blockly/android/ui/FlyoutView.java#L81)):*
> ```Java
>             // TODO (#503): Refactor the action button to be part of the FlyoutCategory.
>             // See Category tabs for an example of binding click listeners to items in a recycler.
>             mActionButton.setTag("createVar");
> ```

Sync with web blockly.  Use the same `<button>` `callbackKey` "CREATE_VARIABLE".

https://github.com/google/blockly/blob/master/core/variables.js#L106

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/FlyoutView.java, line 125 at r2](https://reviewable.io:443/reviews/google/blockly-android/507#-KbqIklfvupH9e8dkEtB:-KbqIklfvupH9e8dkEtC:b-fm6kyb) ([raw file](https://github.com/google/blockly-android/blob/377075ade76b36352eaba49df5efe1cf25426832/blocklylib-core/src/main/java/com/google/blockly/android/ui/FlyoutView.java#L125)):*
> ```Java
> 
>     /**
>      * Sets the alpha used for the background color when the flyout is open. The color will come from the
> ```

line length

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/FlyoutView.java, line 193 at r2](https://reviewable.io:443/reviews/google/blockly-android/507#-KbqIhm0FlY7LbmRYNwK:-KbqIhm0FlY7LbmRYNwL:bbytmq5) ([raw file](https://github.com/google/blockly-android/blob/377075ade76b36352eaba49df5efe1cf25426832/blocklylib-core/src/main/java/com/google/blockly/android/ui/FlyoutView.java#L193)):*
> ```Java
>      * Calculates the workspace point for a {@link PendingDrag}, such that the
>      * {@link MotionEvent#ACTION_DOWN} location remains in the same location on the screen
>      * (i.e., under the user's finger), and calls {@link BlockListView.OnDragListBlock#getDraggableBlockGroup}
> ```

Line length

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-android/507)*
<!-- Sent from Reviewable.io -->

RoboErikG commented:
per discussion, split functionality between the fragment and a helper class that sets up the recyclerview.

---

Review status: 12 of 38 files reviewed at latest revision, 18 unresolved discussions.

---

*[blocklydemo/src/main/java/com/google/blockly/android/demo/StylesActivity.java, line 32 at r1](https://reviewable.io:443/reviews/google/blockly-android/507#-KbWaAbbKSw8hdznfMQ1:-Kbuz4kDGI_EDnyjmtix:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/4ed5c0a35d55fd7c462c8b798f6863f293b4af8e/blocklydemo/src/main/java/com/google/blockly/android/demo/StylesActivity.java#L32)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Fixed fully in #508. Revert if you want a simpler merge.
</blockquote></details>

Done.

---

*[blocklydemo/src/main/java/com/google/blockly/android/demo/StylesActivity.java, line 79 at r1](https://reviewable.io:443/reviews/google/blockly-android/507#-KbWau1sPnoxCyAndsKe:-Kbv1RWQ622v1cJS2CuU:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/4ed5c0a35d55fd7c462c8b798f6863f293b4af8e/blocklydemo/src/main/java/com/google/blockly/android/demo/StylesActivity.java#L79)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Is there a particular reason to add this to the styling demo?  If you're just demoing a different layout, can we use a new activity, named to describe why it is different.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-android/507#-:-Kbv1a0VZVD-fGkuV9jV:bmwoks4)*
<!-- Sent from Reviewable.io -->

RoboErikG commented:
And cleaned up remaining comments.

---

Review status: 9 of 43 files reviewed at latest revision, 18 unresolved discussions.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java, line 65 at r1](https://reviewable.io:443/reviews/google/blockly-android/507#-KbWwBzysr-401fTZgRs:-Kbv2RaKAQR1yPYcMqjm:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/4ed5c0a35d55fd7c462c8b798f6863f293b4af8e/blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java#L65)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

TODO
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java, line 116 at r1](https://reviewable.io:443/reviews/google/blockly-android/507#-KbWwbiVeZHd8cOrWT68:-Kbv2XyEBQLTmCaL6kd8:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/4ed5c0a35d55fd7c462c8b798f6863f293b4af8e/blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java#L116)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

mToolboxFlyoutFragment, or just mToolboxFlyout, to distinguish it from the other FlyoutFragment (mTrashFragment).
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java, line 853 at r1](https://reviewable.io:443/reviews/google/blockly-android/507#-KbWxXwPAipvlc5ZcvhV:-Kbv3-2XSLNPW8Fg6_Gp:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/4ed5c0a35d55fd7c462c8b798f6863f293b4af8e/blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java#L853)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Rename as onBackToCloseFlyouts()
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java, line 319 at r2](https://reviewable.io:443/reviews/google/blockly-android/507#-KbkJp1FkzylP_zKlDx7:-Kbv2bkVnLQKFGpXqVf8:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/377075ade76b36352eaba49df5efe1cf25426832/blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java#L319)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

TODO
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/CategoryFragment.java, line 42 at r1](https://reviewable.io:443/reviews/google/blockly-android/507#-KbWxkGi4dmxwCEgeut4:-Kbv4A9EGD023KpaT-1x:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/4ed5c0a35d55fd7c462c8b798f6863f293b4af8e/blocklylib-core/src/main/java/com/google/blockly/android/CategoryFragment.java#L42)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

CategoriesFragment, or CategorySelectorFragment?
TODO(###): Define an interface that app developers can implement their own UI.
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/FlyoutFragment.java, line 242 at r1](https://reviewable.io:443/reviews/google/blockly-android/507#-KbXWySgy0T55SALVtb6:-Kbv4lL8kstzvhxOIj1D:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/4ed5c0a35d55fd7c462c8b798f6863f293b4af8e/blocklylib-core/src/main/java/com/google/blockly/android/FlyoutFragment.java#L242)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

This comment is obsolete.  Please remove.
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/FlyoutFragment.java, line 131 at r2](https://reviewable.io:443/reviews/google/blockly-android/507#-KbqUQhN2hhfn8ebRrjY:-Kbv4LyYYTBYiU-lK8e-:b-s637pw) ([raw file](https://github.com/google/blockly-android/blob/377075ade76b36352eaba49df5efe1cf25426832/blocklylib-core/src/main/java/com/google/blockly/android/FlyoutFragment.java#L131)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Maybe not now, but...
I think this need a FrameView or similar, whose contents can be replaced with the correct view in the case of `custom` categories.
At least add a TODO.
</blockquote></details>

I'm not sure I understand. This view is independent of the view used for categories. However, I did notice I wasn't using any RelativeLayout methods so I've made this just a View.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyController.java, line 262 at r2](https://reviewable.io:443/reviews/google/blockly-android/507#-KbkKODwAPWxlASsemX8:-Kbv4zES_UjlJQ3_dL_z:bq372jf) ([raw file](https://github.com/google/blockly-android/blob/377075ade76b36352eaba49df5efe1cf25426832/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyController.java#L262)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Based on a comment above, I think you were considering breaking this into two set methods, which I support.
</blockquote></details>

Which comment? I did consider that but it's a lot easier to do the configuration if you have both of these pieces at the same time and don't have to worry about the order they're set in.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/FlyoutController.java, line 33 at r2](https://reviewable.io:443/reviews/google/blockly-android/507#-KbkL9mJeDjBPaCMXM57:-Kbv5NEV_YvnXsFaERBD:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/377075ade76b36352eaba49df5efe1cf25426832/blocklylib-core/src/main/java/com/google/blockly/android/control/FlyoutController.java#L33)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Add: Ensures the toolbox and trash block-list flyouts are populated, opened, and closed in coordination.
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/FlyoutController.java, line 173 at r2](https://reviewable.io:443/reviews/google/blockly-android/507#-KbqJWHmAwTf_xmAg1Q6:-Kbv5ZI0ZxPYjmKBKQv-:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/377075ade76b36352eaba49df5efe1cf25426832/blocklylib-core/src/main/java/com/google/blockly/android/control/FlyoutController.java#L173)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

TODO... Not sure if you intend to address this in this pass.  If not, please clarify "reset fragments"
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/FlyoutController.java, line 233 at r2](https://reviewable.io:443/reviews/google/blockly-android/507#-KbkSL9t_wvS4MrIPH18:-Kbv5v8iNMQHsDQJw0FD:bdmmtc0) ([raw file](https://github.com/google/blockly-android/blob/377075ade76b36352eaba49df5efe1cf25426832/blocklylib-core/src/main/java/com/google/blockly/android/control/FlyoutController.java#L233)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

I don't think registering the trashIcon should change the flyout state.
</blockquote></details>

Done. Moved the close to when the trash fragment is set (it should start closed if closable).

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/FlyoutView.java, line 79 at r2](https://reviewable.io:443/reviews/google/blockly-android/507#-KbqL9da10_f5MyLbdUz:-Kbv6MUDo8nJ0VAPV2l-:btqqu5l) ([raw file](https://github.com/google/blockly-android/blob/377075ade76b36352eaba49df5efe1cf25426832/blocklylib-core/src/main/java/com/google/blockly/android/ui/FlyoutView.java#L79)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

I'm assuming moving this variable-specific code into a variable class
</blockquote></details>

Correct.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/FlyoutView.java, line 81 at r2](https://reviewable.io:443/reviews/google/blockly-android/507#-KbqKCwSJz4yOfohBUCZ:-Kbv7-vPpcEHIh1gFgXy:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/377075ade76b36352eaba49df5efe1cf25426832/blocklylib-core/src/main/java/com/google/blockly/android/ui/FlyoutView.java#L81)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Sync with web blockly.  Use the same `<button>` `callbackKey` "CREATE_VARIABLE".

https://github.com/google/blockly/blob/master/core/variables.js#L106
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/FlyoutView.java, line 125 at r2](https://reviewable.io:443/reviews/google/blockly-android/507#-KbqIklfvupH9e8dkEtB:-Kbv70znBOk3nSFtqmhR:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/377075ade76b36352eaba49df5efe1cf25426832/blocklylib-core/src/main/java/com/google/blockly/android/ui/FlyoutView.java#L125)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

line length
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/FlyoutView.java, line 193 at r2](https://reviewable.io:443/reviews/google/blockly-android/507#-KbqIhm0FlY7LbmRYNwK:-Kbv7EDaPmyDF34VOEgR:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/377075ade76b36352eaba49df5efe1cf25426832/blocklylib-core/src/main/java/com/google/blockly/android/ui/FlyoutView.java#L193)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Line length
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-android/507#-:-Kbv7UgkuCJv9nkt952T:blqe8cx)*
<!-- Sent from Reviewable.io -->

AnmAtAnm commented:
Reviewed 17 of 39 files at r1, 1 of 1 files at r2, 20 of 20 files at r3.
Review status: all files reviewed at latest revision, 16 unresolved discussions.

---

*[blocklydemo/src/main/java/com/google/blockly/android/demo/StyledFlyoutsActivity.java, line 32 at r3](https://reviewable.io:443/reviews/google/blockly-android/507#-KbvQ5tKMAdU9ylASgJy:-KbvQ5tKMAdU9ylASgJz:bbd87bs) ([raw file](https://github.com/google/blockly-android/blob/bf518d0dc4fbed77cce5690340d241ea9f506cbe/blocklydemo/src/main/java/com/google/blockly/android/demo/StyledFlyoutsActivity.java#L32)):*
> ```Java
>  *
>  */
> public class StyledFlyoutsActivity extends AbstractBlocklyActivity {
> ```

Add TODO: Add menu items to demonstrate alternate layouts (restart activity with intent params?).

---

*[blocklydemo/src/main/res/layout/styled_content.xml, line 2 at r3](https://reviewable.io:443/reviews/google/blockly-android/507#-KbvJMlPzPd3utOKEWIO:-KbvJMlPzPd3utOKEWIP:bbpcxxe) ([raw file](https://github.com/google/blockly-android/blob/bf518d0dc4fbed77cce5690340d241ea9f506cbe/blocklydemo/src/main/res/layout/styled_content.xml#L2)):*
> ```XML
> <?xml version="1.0" encoding="utf-8"?>
> <RelativeLayout
> ```

Worth adding a comment about how this differs from blockly_unified)workspace.xml

---

*[blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java, line 67 at r3](https://reviewable.io:443/reviews/google/blockly-android/507#-KbvQmJkF9Vhr309RkXW:-KbvQmJkF9Vhr309RkXX:bbytmq5) ([raw file](https://github.com/google/blockly-android/blob/bf518d0dc4fbed77cce5690340d241ea9f506cbe/blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java#L67)):*
> ```Java
>  * {@link ActionBar} can be replaced by overriding {@link #onCreateContentView}.  After
>  * {@link #onCreateContentView}, the base implementation of {@link #onCreateFragments()} looks for
>  * the {@link WorkspaceFragment}, the toolbox's {@link FlyoutFragment} and {@link CategorySelectorFragment},
> ```

Line length

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyController.java, line 262 at r2](https://reviewable.io:443/reviews/google/blockly-android/507#-KbkKODwAPWxlASsemX8:-KbvKfcrlT4LEI-pOvNA:b-qc15d4) ([raw file](https://github.com/google/blockly-android/blob/377075ade76b36352eaba49df5efe1cf25426832/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyController.java#L262)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

Which comment? I did consider that but it's a lot easier to do the configuration if you have both of these pieces at the same time and don't have to worry about the order they're set in.
</blockquote></details>

In AbstractBlocklyActivity.onCreate() you had a `// TODO` next to `.setToolboxFragment(mToolboxFlyoutFragment, mCategoryFragment);`
I'm fine with this.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/BlockDrawerFragment.java, line 28 at r3](https://reviewable.io:443/reviews/google/blockly-android/507#-KbvOEW3ve6jWpSJiE-t:-KbvOEW3ve6jWpSJiE-u:b4o3121) ([raw file](https://github.com/google/blockly-android/blob/bf518d0dc4fbed77cce5690340d241ea9f506cbe/blocklylib-core/src/main/java/com/google/blockly/android/ui/BlockDrawerFragment.java#L28)):*
> ```Java
> 
> /**
>  * TODO: DELETE THIS CLASS
> ```

Is this still used?

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/BlockListView.java, line 45 at r3](https://reviewable.io:443/reviews/google/blockly-android/507#-KbvONPDX5ys7zhyWyYk:-KbvONPDX5ys7zhyWyYl:b-8cjogr) ([raw file](https://github.com/google/blockly-android/blob/bf518d0dc4fbed77cce5690340d241ea9f506cbe/blocklylib-core/src/main/java/com/google/blockly/android/ui/BlockListView.java#L45)):*
> ```Java
>  * define a {@link RecyclerView.LayoutManager}.
>  */
> public class BlockListView extends RecyclerView {
> ```

Still used?

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/CategoryTabs.java, line 250 at r3](https://reviewable.io:443/reviews/google/blockly-android/507#-KbvReVGzVruBOW195b7:-KbvReVGzVruBOW195b8:bq2ex1y) ([raw file](https://github.com/google/blockly-android/blob/bf518d0dc4fbed77cce5690340d241ea9f506cbe/blocklylib-core/src/main/java/com/google/blockly/android/ui/CategoryTabs.java#L250)):*
> ```Java
>      * Callback for when the user clicks on a category.
>      */
>     public abstract static class Callback {
> ```

Move back to the top of the class, since it is part of the public interface.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/FlyoutCallback.java, line 25 at r3](https://reviewable.io:443/reviews/google/blockly-android/507#-KbvMtT1RQYetNjlFCqS:-KbvMtT1RQYetNjlFCqT:big4ydn) ([raw file](https://github.com/google/blockly-android/blob/bf518d0dc4fbed77cce5690340d241ea9f506cbe/blocklylib-core/src/main/java/com/google/blockly/android/ui/FlyoutCallback.java#L25)):*
> ```Java
> 
> /**
>  *
> ```

Class doc

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/FlyoutCallback.java, line 32 at r3](https://reviewable.io:443/reviews/google/blockly-android/507#-KbvMUp9CeKFn2FLhg37:-KbvMUp9CeKFn2FLhg38:b-12jy7b) ([raw file](https://github.com/google/blockly-android/blob/bf518d0dc4fbed77cce5690340d241ea9f506cbe/blocklylib-core/src/main/java/com/google/blockly/android/ui/FlyoutCallback.java#L32)):*
> ```Java
>      *
>      * @param v The view that was clicked.
>      * @param action The action tag associated with the view.
> ```

"The callback key associated with the clicked button."
(Trying to keep the same terminology used in the toolbox XML.)

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/FlyoutCallback.java, line 35 at r3](https://reviewable.io:443/reviews/google/blockly-android/507#-KbvMklCw62Xh8zT7XKB:-KbvMklCw62Xh8zT7XKC:b-1u6nle) ([raw file](https://github.com/google/blockly-android/blob/bf518d0dc4fbed77cce5690340d241ea9f506cbe/blocklylib-core/src/main/java/com/google/blockly/android/ui/FlyoutCallback.java#L35)):*
> ```Java
>      * @param category The category that this action was in.
>      */
>     public abstract void onActionClicked(View v, String action, FlyoutCategory category);
> ```

onButtonClicked?

---

*[blocklylib-core/src/main/java/com/google/blockly/model/FlyoutCategory.java, line 34 at r3](https://reviewable.io:443/reviews/google/blockly-android/507#-KbvPIsFfZpHkuvmJkM2:-KbvPIsFfZpHkuvmJkM3:b-eaass7) ([raw file](https://github.com/google/blockly-android/blob/bf518d0dc4fbed77cce5690340d241ea9f506cbe/blocklylib-core/src/main/java/com/google/blockly/model/FlyoutCategory.java#L34)):*
> ```Java
> 
> /**
>  * A category of a toolbox, which holds zero or more blocks and zero or more subcategories.
> ```

"...which holds one or more subcategories _or_ zero or more blocks. Not both."

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-android/507)*
<!-- Sent from Reviewable.io -->

RoboErikG commented:
Review status: 33 of 43 files reviewed at latest revision, 15 unresolved discussions.

---

*[blocklydemo/src/main/java/com/google/blockly/android/demo/StyledFlyoutsActivity.java, line 32 at r3](https://reviewable.io:443/reviews/google/blockly-android/507#-KbvQ5tKMAdU9ylASgJy:-KbvWAt05nR4FOsZFbx5:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/bf518d0dc4fbed77cce5690340d241ea9f506cbe/blocklydemo/src/main/java/com/google/blockly/android/demo/StyledFlyoutsActivity.java#L32)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Add TODO: Add menu items to demonstrate alternate layouts (restart activity with intent params?).
</blockquote></details>

Done.

---

*[blocklydemo/src/main/res/layout/styled_content.xml, line 2 at r3](https://reviewable.io:443/reviews/google/blockly-android/507#-KbvJMlPzPd3utOKEWIO:-KbvWVRzqPzvsuO8ra_-:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/bf518d0dc4fbed77cce5690340d241ea9f506cbe/blocklydemo/src/main/res/layout/styled_content.xml#L2)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Worth adding a comment about how this differs from blockly_unified)workspace.xml
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java, line 67 at r3](https://reviewable.io:443/reviews/google/blockly-android/507#-KbvQmJkF9Vhr309RkXW:-KbvWg91qJhVhmWHU5jJ:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/bf518d0dc4fbed77cce5690340d241ea9f506cbe/blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java#L67)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Line length
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/BlockDrawerFragment.java, line 28 at r3](https://reviewable.io:443/reviews/google/blockly-android/507#-KbvOEW3ve6jWpSJiE-t:-KbvWstW1KF5ZUqWyp4g:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/bf518d0dc4fbed77cce5690340d241ea9f506cbe/blocklylib-core/src/main/java/com/google/blockly/android/ui/BlockDrawerFragment.java#L28)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Is this still used?
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/BlockListView.java, line 45 at r3](https://reviewable.io:443/reviews/google/blockly-android/507#-KbvONPDX5ys7zhyWyYk:-KbvWy7odBb-h84ohb4H:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/bf518d0dc4fbed77cce5690340d241ea9f506cbe/blocklylib-core/src/main/java/com/google/blockly/android/ui/BlockListView.java#L45)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Still used?
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/CategoryTabs.java, line 250 at r3](https://reviewable.io:443/reviews/google/blockly-android/507#-KbvReVGzVruBOW195b7:-KbvXW28wFnkEDJxA0ho:bnbakzy) ([raw file](https://github.com/google/blockly-android/blob/bf518d0dc4fbed77cce5690340d241ea9f506cbe/blocklylib-core/src/main/java/com/google/blockly/android/ui/CategoryTabs.java#L250)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Move back to the top of the class, since it is part of the public interface.
</blockquote></details>

Our style guide says static inner classes go at the very bottom, after non-static inner classes. Moved all the static classes below non-statics.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/FlyoutCallback.java, line 25 at r3](https://reviewable.io:443/reviews/google/blockly-android/507#-KbvMtT1RQYetNjlFCqS:-KbvZ3SEeEPnyCYGQIg-:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/bf518d0dc4fbed77cce5690340d241ea9f506cbe/blocklylib-core/src/main/java/com/google/blockly/android/ui/FlyoutCallback.java#L25)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Class doc
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/FlyoutCallback.java, line 32 at r3](https://reviewable.io:443/reviews/google/blockly-android/507#-KbvMUp9CeKFn2FLhg37:-KbvZ8IKe0gWzW0tYIIO:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/bf518d0dc4fbed77cce5690340d241ea9f506cbe/blocklylib-core/src/main/java/com/google/blockly/android/ui/FlyoutCallback.java#L32)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

"The callback key associated with the clicked button."
(Trying to keep the same terminology used in the toolbox XML.)
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/FlyoutCallback.java, line 35 at r3](https://reviewable.io:443/reviews/google/blockly-android/507#-KbvMklCw62Xh8zT7XKB:-KbvZBQsNPabWw4G7gs3:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/bf518d0dc4fbed77cce5690340d241ea9f506cbe/blocklylib-core/src/main/java/com/google/blockly/android/ui/FlyoutCallback.java#L35)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

onButtonClicked?
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/model/FlyoutCategory.java, line 34 at r3](https://reviewable.io:443/reviews/google/blockly-android/507#-KbvPIsFfZpHkuvmJkM2:-KbvZLlOZYjo1Syzim2Y:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/bf518d0dc4fbed77cce5690340d241ea9f506cbe/blocklylib-core/src/main/java/com/google/blockly/model/FlyoutCategory.java#L34)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

"...which holds one or more subcategories _or_ zero or more blocks. Not both."
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/FlyoutView.java, line 34 at r2](https://reviewable.io:443/reviews/google/blockly-android/507#-KbqT5fXMMgcCspdwnHQ:-KbvZRIZEMlc5bKNShPk:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/377075ade76b36352eaba49df5efe1cf25426832/blocklylib-core/src/main/java/com/google/blockly/android/ui/FlyoutView.java#L34)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Extract RecyclerView initialization and adapter into a helper or factory method. Maybe into the FlyoutController. Aim to make this class as minimal as possible, possibly with the expectation that it may go away / become a simple RelativeLayout once #503 is resolved.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-android/507)*
<!-- Sent from Reviewable.io -->

AnmAtAnm commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 10 of 10 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-android/507#-:-KbvvKqkZIkvQYww1oOF:bnfp4nl)*
<!-- Sent from Reviewable.io -->

AnmAtAnm commented:
There is still a reference to BlockListView in OnDragToTrashListenerTest.

AnmAtAnm commented:
Also crashes if flyout, such as trash, is not present.

RoboErikG commented:
Fixed both and updated the flyouts demo to not use a trash flyout.
